### PR TITLE
Add `verify_preflight_check` for script sequences

### LINKED ISSUE #6368: `forge script --verify` skips verification silently if etherscan API key is not provided
### Component

Forge

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

_No response_

### What command(s) is the bug in?

_No response_

### Operating System

None

### Describe the bug

When running script with `--verify` flag without providing etherscan API key in either foundry.toml or CLI parameter, verification will be silently skipped. This may be unexpected for user and leave him in uncomfortable position of spending gas on deployment of contracts which were not verified.

This happens because of the following snippet: https://github.com/foundry-rs/foundry/blob/3bab117ec76dc1cd63aa6b8ca25eaaef909f2535/crates/forge/bin/cmd/script/sequence.rs#L258-L262

`verify.etherscan.key` is set in `set_chain` function called.

https://github.com/foundry-rs/foundry/blob/3bab117ec76dc1cd63aa6b8ca25eaaef909f2535/crates/forge/bin/cmd/script/verify.rs#L58-L63

`get_etherscan_api_key` will return `None` for keys which are not found in `foundry.toml`, env variable or CLI param

On the other hand, `forge verify` fails with `ETHERSCAN_API_KEY must be set`, if API key is not found

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Closes #6368

## Solution

Adds `ScriptSequence::verify_preflight_check` method which is being called between on-chain simulation and actual transactions broadcasting in cases when `self.verify == true`.

The same naming is used in `forge create` `CreateArgs::verify_preflight_check` so it might make sense to change it, but I didn't figure out another suitable name for it.

This might break some scenarios when `--verify` flag is used for multichain deployments when only subset of chains support etherscan verification.

### CODE REVIEW

File: crates/forge/bin/cmd/script/sequence.rs [hunk 1]
```diff
@@ -246,6 +246,17 @@ impl ScriptSequence {
         Ok((broadcast, cache))
     }
 
+    /// Checks that there is an Etherscan key for the chain id of this sequence.
+    pub fn verify_preflight_check(&self, config: &Config, verify: &VerifyBundle) -> Result<()> {
+        if config.get_etherscan_api_key(Some(self.chain.into())).is_none() &&
+            verify.verifier.verifier == VerificationProviderType::Etherscan
+        {
+            eyre::bail!("Etherscan API key wasn't found for chain id {}", self.chain)
```
Comment (mattsse) [inline crates/forge/bin/cmd/script/sequence.rs:254]: I think this error message should mention that deployment is aborted/not started?

**Review by mattsse — COMMENTED** [review]
this now only deploys with --verify and all used chains have a valid etherscan API key.

I think this is reasonable.

wdyt @mds1

**Comment by klkvr** [comment]
@mattsse did CI failed because of my code? I don't understand what exactly have happened there

**Comment by mds1** [comment]
Previous behavior: If no etherscan API key was set but user specified to verify, verification was silently skipped and everything else proceeds as normal

New behavior: If no etherscan API key was set but user specified to verify, forge will exit with an error after the simulation but before broadcasting.

Is that correct? If so sgtm, but may be better UX to fail fast before simulation?

**Comment by klkvr** [comment]
@mds1 Yes, that's correct

To make it fail before simulation the implementation would have to be less cleaner, because the logic will be different for newly created and resumed scripts. I would be happy to improve it that way, but considering that on-chain simulation usually doesn't take much time (usually less than script execution which will run anyway) it may not be really needed.

btw I have been testing it for multichain scripts and #5047 seems to be the case. Is there any working example of `--resume --multi`? I am always getting panic `to have been filled with a proper rpc`

**Comment by klkvr** [comment]
Also, in case when script fails because of missing etherscan key, transactions are getting saved into `./broadcast` and can be restarted via `--resume`, thus skipping both simulation and execution which wouldn't be possible if it failed before simulation.

**Review by Evalir — APPROVED** [review]
lgtm—i think it's reasonable to include this. cc @mattsse for merging unless we haven't settled on the ux discussion on when to fail?

**Comment by mds1** [comment]
> To make it fail before simulation the implementation would have to be less cleaner, because the logic will be different for newly created and resumed scripts. I would be happy to improve it that way, but considering that on-chain simulation usually doesn't take much time (usually less than script execution which will run anyway) it may not be really needed.

Got it, this seems ok to me then. @Evalir @mattsse if scripting ever gets refactored moving this to fail earlier is a nice UX change to implement if you want to track that somewhere (I've definitely had some scripts that can take 30+ seconds to simulate due to lots of RPC calls)

**Pull request merged**